src/ecs/resources/input.rs

```diff
@@ -62,160 +181,335 @@ impl<'a> Iterator for PressedMouseButtons<'a> {
-        if !self.pressed_mouse_buttons.contains_key(&button) {
-            return false;
+    /// this method can return None, either if no mouse is connected, or if no mouse events have
+    /// been recorded
+    pub fn mouse_position(&self) -> Option<(i32, i32)> {
+        self.current_frame.mouse_position
+    }
+
+    /// Gets the change in position since the last frame.
+    pub fn mouse_position_change(&self) -> (i32, i32) {
+        match (self.current_frame.mouse_position, self.previous_frame.mouse_position) {
+            (Some(current), Some(previous)) => (current.0 - previous.0, current.1 - previous.1),
+            _ => (0, 0),
+        }
+    }
+
+    /// Returns an iterator over the buttons that are currently pressed
+    pub fn pressed_buttons(&self) -> PressedButtons {
+        self.current_frame.pressed_buttons()
+    }
+
+    /// Checks if the given button is currently pressed.
+    pub fn button_pressed(&self, button: Button) -> bool {
+        self.current_frame.button_pressed(button)
+    }
+
+    /// Checks if all the given buttons are pressed.
+    pub fn buttons_pressed(&self, buttons: &[Button]) -> bool {
+        self.current_frame.buttons_pressed(buttons)
+    }
+
+    /// Checks if the given button was pressed on this frame.
+    pub fn button_down(&self, button: Button) -> bool {
+        match button {
+            Button::Key(k) => self.key_down(k),
+            Button::Mouse(b) => self.mouse_button_down(b),
         }
-        if let Some(value) = self.pressed_mouse_buttons.get_mut(&button) {
-            if *value == KeyQueryState::NotQueried {
-                *value = KeyQueryState::Queried;
-                return true;
+    }
+
+    /// Checks if the given button was released on this frame.
+    pub fn button_released(&self, button: Button) -> bool {
+        match button {
+            Button::Key(k) => self.key_released(k),
+            Button::Mouse(b) => self.mouse_button_released(b),
+        }
+    }
+
+    /// Checks if the all the given buttons are being pressed and at least one was pressed this frame.
+    pub fn buttons_down(&self, buttons: &[Button]) -> bool {
+        buttons.iter().any(|&b| self.button_down(b)) &&
+        buttons.iter().all(|&b| self.button_pressed(b))
+    }
+
+    /// Returns the value of an axis by the i32 id, if the id doesn't exist this returns None.
+    pub fn axis_value(&self, id: i32) -> Option<f32> {
+        self.axes
+            .get(&id)
+            .map(|a| {
+                let pos = self.button_down(a.pos);
+                let neg = self.button_down(a.neg);
+                if pos == neg {
+                    0.0
+                } else if pos {
+                    1.0
+                } else {
+                    -1.0
+                }
+            })
+    }
+
+    // Pressed actions has been omitted as the function is a little difficult to write and I
+    // can't think of a use case for it.
```

Comment:
I just renamed the functions to
pressed_keys()
key_is_pressed()
keys_are_pressed()

I think that helps disambiguate it a bit better.